trpcultivate_phenotypes/src/Service/TripalCultivatePhenotypesTraitsService.php

```diff
@@ -255,163 +308,286 @@ public function insertTrait($trait, $schema = NULL) {
+    if (!$method_rec) {
+      return 0;
+    }
+
+    // Unit.
+    $unit_rec = $this->getTraitAsset($unit, 'unit');
+    if (!$unit_rec) {
       return 0;
     }
 
-    $sql = "SELECT value FROM {1:cvtermprop} WHERE cvterm_id = :c_id AND type_id = :t_id LIMIT 1";
-    $data_type = $this->chado->query($sql, [':c_id' => $unit_id, ':t_id' => $this->terms['unit_type']])
-      ->fetchField();
+    // Append unit data type to the unit data object.
+    $unit_id = $unit_rec->cvterm_id;
+    $unit_data_type = $this->getMethodUnitDataType($unit_id);
+    $unit_rec->{'data_type'} = $unit_data_type;
+
+    return [
+      'trait' => $trait_rec,
+      'method' => $method_rec,
+      'unit'  => $unit_rec
+    ];
+  }
+
+  /**
+   * Get trait asset - trait, method or unit.
+   *
+   * @param string|int $key
+   *   A string value is the name (cvterm.name), whereas an integer value
+   *   is the id number (cvterm.cvterm_id).
+   * @param string $type
+   *   trait, method or unit asset type. Trait is the default.
+   *
+   * @return object
+   *   Trait asset record object or 0 if not found.
+   */
+  public function getTraitAsset($key, $type = 'trait') {
+    // Configuration check.
+    $genus_config = $this->config;
+    if (!$genus_config) {
+      // Genus not set.
+      throw new \Exception(t('No genus has been set. See setting a genus in the Traits Service and make sure to
+        use a configured genus. To configure a genus or see all configured genus, go to @url.',
+        ['@url' => Url::fromRoute('trpcultivate_phenotypes.settings_ontology')->toString()]
+      ));
+    }
+
+    // Parameter check.
+    if (!in_array($type, ['trait', 'method', 'unit'])) {
+      // Not a valid parameter asset type value..
+      throw new \Exception(t('Not a valid trait asset type value provided. Trait asset getter expects type to be the
+        string trait, method or unit.'
+      ));
+    }
+
+    if (empty($key) || (is_numeric($key) && (int) $key < 0)) {
+      // Not a valid asset key value (0, negative values or an empty string).
+      throw new \Exception(t('Not a valid @type key value provided. The trait asset getter expects a string name or
+        an integer id key value.', ['@type' => $type]
+      ));
+    }
+
+
+    // Query trait asset.
+    if (is_numeric($key)) {
+      // Asset parameter key is the id number.
+      $asset_rec = $this->chado->query(
+        "SELECT * FROM {1:cvterm} WHERE cvterm_id = :value",
+        [':value' => (int) $key]
+      )
+        ->fetchAll();
+
+      if ($asset_rec && (int) $asset_rec[0]->cv_id != $genus_config[ $type ]['id']) {
+        // The id number seems to be of a trait asset that is outside of
+        // the cv: asset type (trait, method or unit) the genus was configured.
+        throw new \Exception(t('The requested trait asset @type : id @key CV value does not match the CV the genus was configured.',
+          ['@type' => $type, '@key' => $key]
+        ));
+      }
+    }
+    else {
```

Comment:
We had discussed first assuming an integer was a cvterm but if no cvterm had that id then treating the integer like a potential name. This block doesn't support that right now.